sources/web/datalab/jupyter.ts

```diff
@@ -81,67 +88,78 @@ function pipeOutput(stream: NodeJS.ReadableStream, port: number, error: boolean)
  * Starts the Jupyter server, and then creates a proxy object enabling
  * routing HTTP and WebSocket requests to Jupyter.
  */
-function createJupyterServer(userId: string): JupyterServer {
+function createJupyterServer(userId: string, resolved: (server: JupyterServer)=>void, rejected:()=>void) {
```

Comment:
The rejected logic can be a loop inside this function, which seems easier to understand compared to recursive calls.